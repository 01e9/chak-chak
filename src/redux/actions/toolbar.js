export const ACTION_TOOLBAR_POSITION_SET = 'TOOLBAR_POSITION';

export const actionToolbarPositionSet = (x, y) => ({
    type: ACTION_TOOLBAR_POSITION_SET,
    x,
    y
})
