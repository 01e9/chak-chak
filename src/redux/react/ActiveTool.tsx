import * as React from "react";
import { TOOL_NAME_BRUSH } from "@/constants/tools";
import { connect } from "react-redux";
import { IState } from "@/redux/reducers";

interface IInternalProps {
    activeTool: string;
}

const mapStateToProps = ({ activeTool }: IState) => ({ activeTool });

const render = ({ activeTool }: IInternalProps) => {
    switch (activeTool) {
        case TOOL_NAME_BRUSH: return <div>BRUSH</div>;
        default: return null;
    }
};

export default connect(mapStateToProps)(render);
