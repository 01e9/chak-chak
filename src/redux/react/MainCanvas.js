import * as React from 'react'
import Canvas from "@/react/Canvas";
import { connect } from 'react-redux'
import selectMainImage from "@/redux/selectors/mainImage";
import { ImageStorageConsumer } from "@/react/context/ImageStorage";

const mapStateToProps = (state, ownProps) => ({
    image: selectMainImage(state, ownProps.imageStorage)
})

const MainCanvasConnected = connect(mapStateToProps, null)(Canvas)

const MainCanvas = () => (
    <ImageStorageConsumer>
        {imageStorage => <MainCanvasConnected imageStorage={imageStorage} />}
    </ImageStorageConsumer>
)

export default MainCanvas
