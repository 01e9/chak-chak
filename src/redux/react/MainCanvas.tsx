import * as React from 'react'
import Canvas, { ICanvasProps } from "@/react/Canvas";
import { connect } from 'react-redux'
import selectMainImage from "@/redux/selectors/mainImage";
import { ImageStorageConsumer } from "@/react/context/ImageStorage";
import { IState } from "@/redux/reducers";
import ImageStorage from "@/storage/ImageStorage";

interface IMainCanvasConnectedProps {
    imageStorage: ImageStorage;
}

const mapStateToProps = (state: IState, ownProps: IMainCanvasConnectedProps): (
    Pick<ICanvasProps, "image" | "aspectRatio">
) => ({
    image: selectMainImage(state, ownProps.imageStorage),
    aspectRatio: state.size.aspectRatio,
})

const MainCanvasConnected = connect(mapStateToProps)(Canvas)

const MainCanvas = () => (
    <ImageStorageConsumer>
        {imageStorage => <MainCanvasConnected imageStorage={imageStorage} />}
    </ImageStorageConsumer>
)

export default MainCanvas
