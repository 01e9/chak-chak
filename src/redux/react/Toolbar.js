import * as React from 'react'
import { connect } from 'react-redux'
import Toolbar_ from "@/react/Toolbar";
import { selectToolbarPosition } from "@/redux/selectors/toolbar";
import { actionToolbarPositionSet } from "@/redux/actions/toolbar";
import ToolBrushControl from "@/redux/react/tool/brush/Control";
import { TOOL_NAME_BRUSH } from "@/constants/tools";

const mapStateToProps = (state, ownProps) => {
    const position = selectToolbarPosition(state);

    return {
        left: position.x,
        top: position.y,
        tools: [
            () => <ToolBrushControl key={TOOL_NAME_BRUSH} />
        ]
    }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
    onPositionUpdate: (left, top) => dispatch(actionToolbarPositionSet(left, top))
})

const Toolbar = connect(mapStateToProps, mapDispatchToProps)(Toolbar_)

export default Toolbar
