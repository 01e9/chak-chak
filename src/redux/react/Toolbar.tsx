import * as React from 'react'
import { connect } from 'react-redux'
import Toolbar_, { IToolbarProps } from "@/react/Toolbar";
import { selectToolbarPosition } from "@/redux/selectors/toolbar";
import { actionToolbarPositionSet } from "@/redux/actions/toolbar";
import ToolBrushControl from "@/redux/react/tool/brush/Control";
import { TOOL_NAME_BRUSH } from "@/constants/tools";
import { IState } from "@/redux/reducers";
import { IAction } from "@/redux/utils/actions";
import { Dispatch } from "redux";

const mapStateToProps = (state: IState): Omit<IToolbarProps, "onPositionUpdate"> => {
    const position = selectToolbarPosition(state);

    return {
        left: position.x,
        top: position.y,
        tools: [
            {key: TOOL_NAME_BRUSH, Component: ToolBrushControl}
        ]
    }
}

const mapDispatchToProps = (dispatch: Dispatch<IAction>): Pick<IToolbarProps, "onPositionUpdate"> => ({
    onPositionUpdate: (x, y) => dispatch(actionToolbarPositionSet({x, y}))
})

const Toolbar = connect(mapStateToProps, mapDispatchToProps)(Toolbar_)

export default Toolbar
