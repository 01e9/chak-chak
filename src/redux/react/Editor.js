import Editor_ from "@/react/Editor";
import { connect } from 'react-redux'
import MainCanvas from "@/redux/react/MainCanvas";
import Toolbar from "@/redux/react/Toolbar";

const mapStateToProps = (state, ownProps) => ({
    Canvas: MainCanvas,
    Toolbar: Toolbar,
    ActiveTool: () => '[ActiveTool]' // fixme
})

const Editor = connect(mapStateToProps, null)(Editor_)

export default Editor
