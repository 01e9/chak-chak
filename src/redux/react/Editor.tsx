import * as React from "react";
import Editor_ from "@/react/Editor";
import { connect } from 'react-redux'
import MainCanvas from "@/redux/react/MainCanvas";
import Toolbar from "@/redux/react/Toolbar";
import { IState } from "@/redux/reducers";

const mapStateToProps = (state: IState) => ({
    Canvas: MainCanvas,
    Toolbar: Toolbar,
    ActiveTool: () => <span>[ActiveTool]</span> // fixme
})

const Editor = connect(mapStateToProps)(Editor_)

export default Editor
