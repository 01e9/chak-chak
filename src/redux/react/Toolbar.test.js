import * as React from 'react'
import Toolbar from "@/redux/react/Toolbar";
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import ImageStorage from "@/storage/ImageStorage";
import createReduceRoot from "@/redux/reducers";
import { createMount } from "@material-ui/core/test-utils";
import { actionToolbarPositionSet } from "@/redux/actions/toolbar";

const mockStore = configureStore([])

function createMounted(action = {type: null}) {
    const stateWithToolbarPosition = createReduceRoot(new ImageStorage())(undefined, action)
    const store = mockStore(stateWithToolbarPosition);
    const wrapper = createMount()(
        <Provider store={store}>
            <div><Toolbar/></div>
        </Provider>
    )
    return {wrapper, store}
}

it('snapshot', () => {
    const {wrapper} = createMounted(actionToolbarPositionSet(101, 102));
    expect(wrapper.html()).toMatchSnapshot()
})

describe('mapDispatchToProps', () => {
    it('onPositionUpdate', () => {
        const {wrapper, store} = createMounted(actionToolbarPositionSet(0, 0));

        const dragButton = wrapper.find('.drag-button').hostNodes();
        expect(dragButton).toHaveLength(1);

        expect(store.getActions()).toHaveLength(0);

        dragButton.simulate('gotPointerCapture', {pageX: 1, pageY: 2});
        dragButton.simulate('pointerMove', {pageX: 10, pageY: 10});
        dragButton.simulate('lostPointerCapture', {});

        const actions = store.getActions();
        expect(actions).toHaveLength(1);
        expect(actions[0]).toEqual(actionToolbarPositionSet(9, 8));
    })
})
