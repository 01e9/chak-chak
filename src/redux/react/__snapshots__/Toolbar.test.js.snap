// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `

<div>
  <div class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
       style="top: 102px; left: 101px;"
  >
    <button class="MuiButtonBase-root MuiButton-root undefined drag-button MuiButton-text"
            tabindex="0"
            type="button"
    >
      <span class="MuiButton-label">
        <svg class="MuiSvgIcon-root"
             focusable="false"
             viewbox="0 0 24 24"
             aria-hidden="true"
             role="presentation"
        >
          <path fill="none"
                d="M0 0h24v24H0V0z"
          >
          </path>
          <path d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z">
          </path>
        </svg>
      </span>
    </button>
    <button class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
    >
      <span class="MuiButton-label">
        <svg class="MuiSvgIcon-root"
             focusable="false"
             viewbox="0 0 24 24"
             aria-hidden="true"
             role="presentation"
        >
          <path fill="none"
                d="M0 0h24v24H0z"
          >
          </path>
          <path d="M7 14c-1.66 0-3 1.34-3 3 0 1.31-1.16 2-2 2 .92 1.22 2.49 2 4 2 2.21 0 4-1.79 4-4 0-1.66-1.34-3-3-3zm13.71-9.37l-1.34-1.34a.9959.9959 0 0 0-1.41 0L9 12.25 11.75 15l8.96-8.96c.39-.39.39-1.02 0-1.41z">
          </path>
        </svg>
      </span>
      <span class="MuiTouchRipple-root">
      </span>
    </button>
  </div>
</div>

`;
